nix-exec broken on master, fine on release-16.09

## Issue description

Nix-exec package can't be built on master branch.
Broken for i686 https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.nix-exec.i686-linux/all and x86_64 https://hydra.nixos.org/job/nixos/trunk-combined/nixpkgs.nix-exec.x86_64-linux/all since 2016-12-11

It seems related to a nix langage API change, commit https://github.com/NixOS/nix/commit/cb5e7254b66a06b78a5659551a6f28fc67e52267.

https://hydra.nixos.org/build/44683739/nixlog/1

~~~
/nix/store/add4869c9wrvg611l2py4j3hhb6wd5vp-nix-1.12pre4911_b30d1e7-dev/include/nix/util.hh:172:10: error: 'void nix::AutoCloseFD::close()' is private
     void close();
          ^
src/fetchgit.cc:99:27: error: within this context
       pipe.readSide.close();
~~~

cc @shlevy 

## Technical details

* System: NixOS official hydra & local build with `nix-shell -p nox --run "nox-review wip"`
* Nixpkgs version: 17.03pre99759.f66d782

--- a/src/fetchgit.cc
+++ b/src/fetchgit.cc
@@ -97,14 +97,14 @@ extern "C" void fetchgit( nix::EvalState & state
       throw SysError("forking to run fetchgit");
     case 0:
       pipe.readSide = -1;
-      if (dup2(pipe.writeSide, STDOUT_FILENO) == -1)
+      if (dup2(pipe.writeSide.get(), STDOUT_FILENO) == -1)
         err(214, "duping pipe to stdout");
       /* const-correct, execv doesn't modify it c just has dumb casting rules */
       execv(fetchgit_path, const_cast<char * const *>(argv));
       err(212, "executing %s", fetchgit_path);
   }
   pipe.writeSide = -1;
-  auto path = nix::drainFD(pipe.readSide);
+  auto path = nix::drainFD(pipe.readSide.get());
 
   int status;
   errno = 0;

--- a/src/fetchgit.cc
+++ b/src/fetchgit.cc
@@ -78,7 +78,7 @@ extern "C" void fetchgit( nix::EvalState & state
   auto submodules_iter = args[0]->attrs->find(submodules_sym);
   auto do_submodules = submodules_iter == args[0]->attrs->end() ?
     true :
-    state.forceBool(*submodules_iter->value);
+    state.forceBool(*submodules_iter->value, *submodules_iter->pos);
 
   constexpr char fetchgit_path[] = NIXEXEC_LIBEXEC_DIR "/fetchgit.sh";
   const char * const argv[] = { fetchgit_path

--- a/src/fetchgit.cc
+++ b/src/fetchgit.cc
@@ -96,14 +96,14 @@ extern "C" void fetchgit( nix::EvalState & state
     case -1:
       throw SysError("forking to run fetchgit");
     case 0:
-      pipe.readSide.close();
+      pipe.readSide = -1;
       if (dup2(pipe.writeSide, STDOUT_FILENO) == -1)
         err(214, "duping pipe to stdout");
       /* const-correct, execv doesn't modify it c just has dumb casting rules */
       execv(fetchgit_path, const_cast<char * const *>(argv));
       err(212, "executing %s", fetchgit_path);
   }
-  pipe.writeSide.close();
+  pipe.writeSide = -1;
   auto path = nix::drainFD(pipe.readSide);
 
   int status;

